cpp/src/arrow/meson.build

```diff
@@ -422,6 +422,8 @@ arrow_lib = library(
     include_directories: arrow_includes,
     dependencies: arrow_deps,
     install: true,
+    gnu_symbol_visibility: 'hidden',
+    cpp_shared_args: ['-DARROW_EXPORTING'],
```

Comment:
FWIW its not really necessary to define a macro like this on non-Windows platforms. We could alternatively just simplify the ARROW_EXPORT macro to always use default symbol visibility.

I figure that change is better left to a follow up though, so for now going with CMake feature parity